src/persist/src/operators/stream.rs

```diff
@@ -301,17 +329,183 @@ where
+                        primary_notificator.notify_at(cap.retain());
+                    }
+                });
+
+                // Consume condition input data but throw it away. We only use this
+                // input to track the frontier (to know how far we're sealed up).
+                // TODO: There should be a better way for doing this, maybe?
+                condition_data_input.for_each(|cap, data| {
+                    data.swap(&mut condition_data_buffer);
+                    condition_data_buffer.drain(..);
+
+                    if active_seal_operator {
+                        // Explicitly seal at the time of received data. We also repeatedly set up
+                        // notifies based on the frontier, below, but also using the
+                        // time/capability of incoming data might make things more fine-grained.
+                        condition_notificator.notify_at(cap.retain());
+                    }
+                });
+
+                if !active_seal_operator {
+                    return;
+                }
+
+                (&mut condition_notificator).for_each(|cap, _count, notificator| {
+                    log::trace!(
+                        "In {}, sealing condition input up to {}",
+                        &operator_name,
+                        cap.time(),
+                    );
+
+                    // Notify when the frontier advances again.
+                    let frontier: Vec<_> = notificator.frontier(1).iter().cloned().collect();
+                    for frontier_element in frontier {
+                        notificator.notify_at(cap.delayed(&frontier_element));
+                    }
+
+                    // TODO: Don't block on the seal. Instead, we should yield from the
+                    // operator and/or find some other way to wait for the seal to succeed.
+                    let result = condition_write.seal(*cap.time()).recv();
+
+                    if let Err(e) = result {
+                        let mut session = data_output.session(&cap);
+                        log::error!(
+                            "Error sealing {} (condition) up to {}: {:?}",
+                            &operator_name,
+                            cap.time(),
+                            e
+                        );
+                        // TODO: make error retractable? Probably not...
+                        session.give(Err((e.to_string(), *cap.time(), 1)));
+                    }
+                });
+
+                (&mut primary_notificator).for_each(|cap, _count, notificator| {
+                    log::trace!(
+                        "In {}, sealing primary input up to {}",
+                        &operator_name,
+                        cap.time(),
+                    );
+
+                    // Notify when the frontier advances again.
+                    let frontier: Vec<_> = notificator.frontier(0).iter().cloned().collect();
```

Comment:
why is this 0 when line 436 is `frontier(1)`? are the instances of `notificator` somehow connected?

Reply:
Please see also my answer above: https://github.com/MaterializeInc/materialize/pull/8546#discussion_r723052390.

But yes, there is only one notificator underneath it all, and you can access the frontiers for the inputs separately.